// gen_dep_tag@0.0.3/lib/index.js
import * as fs from "node:fs";
import * as path from "node:path";
import * as process from "node:process";
import { jsdelivr } from "./cdn.js";
import { sri } from "./sri.js";
export * from "./cdn.js";
export * from "./presets.js";
function pkgVersion(pkgName) {
    return pkgJson(pkgName).version || "";
}
function pkgJson(pkgName) {
    const pkgFile = pkgPath(pkgName, "package.json");
    return JSON.parse(fs.readFileSync(pkgFile, "utf8"));
}
function pkgPath(pkgName, ...subPath) {
    const pwd = process.cwd();
    return path.join(pwd, "node_modules", pkgName, ...subPath);
}
function resolveEntry(pkgName) {
    const package_json = pkgJson(pkgName);
    if (package_json.unpkg) {
        return package_json.unpkg;
    }
    if (package_json.jsdelivr) {
        return package_json.jsdelivr;
    }
    if (typeof package_json.browser === "string") {
        return package_json.browser;
    }
    if (package_json.main) {
        return package_json.main;
    }
    throw Error(`Cannot resolve entry for package ${pkgName}`);
}
function tag(packageName, algorithm, filePath) {
    let file_path = filePath || resolveEntry(packageName);
    if (file_path.startsWith("./")) {
        file_path = file_path.slice(2);
    }
    const local_path = pkgPath(packageName, file_path);
    const version = pkgVersion(packageName);
    const integrity = algorithm ? sri(local_path, algorithm) : "";
    return {
        version,
        path: file_path,
        integrity,
    };
}
export function tagBuilder(config) {
    const sri = config?.sri || false;
    const algorithm = sri === true ? "sha256" : sri;
    const cdn = config?.cdn || jsdelivr;
    return (source) => {
        let name, entry;
        if (typeof source === "string") {
            name = source;
        }
        else {
            name = source.name;
            entry = source.entry;
        }
        const { version, path, integrity } = tag(name, algorithm, entry);
        const url = cdn(name, version, path);
        if (algorithm) {
            return { url, integrity };
        }
        else {
            return { url };
        }
    };
}
